### PR TITLE
PHP Target

### LINKED ISSUE #1850: PHP Target
I would like to propose introducing a PHP target and runtime.

Is anyone interested in joining me to implement a PHP target?

### PR DESCRIPTION
**Pull request by marcospassos**
This PR introduces a PHP target for Antlr 4, and finally closes #1850.

The runtime was extensively tested against a test suite with 2000+ tests, and pass all the 350+ tests automatically generated by ANTLR.

## Codegen tweaks
I've tweaked some files related to code generation to overcome PHP limitations, such as lack of support to inner classes. I also fixed some test templates that weren't language agnostic.

## Acknowledgments
Many thanks to @yar3333, who did the initial work and @amorimjuliana, who tracked down and fixed most of the critical issues found.

## Maintenance note
Composer is the most popular dependency manager for PHP, and one of its caveats is that it can only detect the manifest file in the root directory. That's why I included the PHP runtime as a Git submodule. Being a submodule has an advantage: we can release new versions of the runtime independently of ANTLR tool. 

Once this PR gets merged, whenever a new version is released, one should ensure to update the submodule to the latest version. Updating the runtime is simple as running `git submodule update`, and can easily be included as part of the CI workflow.

Fixed #1850

### CODE REVIEW

**Comment by ericvergnaud** [comment]
Will do.

File: runtime-testsuite/resources/org/antlr/v4/test/runtime/templates/Swift.test.stg [hunk 1]
```diff
@@ -96,7 +102,7 @@ ParserPropertyCall(parser, property) ::= <<
 <parser>.<property>
 >>
 
-PositionAdjustingLexer() ::= <<
```
Comment (ericvergnaud) [inline runtime-testsuite/resources/org/antlr/v4/test/runtime/templates/Swift.test.stg]: Why is this gone in Swift but not other targets?

**Comment by ericvergnaud** [comment]
@marcospassos 
To start with, wow, it's great that you've been able to create this new target, in so little time and almost zero support...
As you've noticed (you added stuff to that end) we have a CI to ensure only PRs with a green build make it to the master code base (except on rare occasions where unrelated builds fail due to Travis issues).
It seems this PR does not quite meet this requirement yet.
[Travis build](https://travis-ci.org/antlr/antlr4/builds/584364550?utm_source=github_status&utm_medium=notification)
The PHP tests don't pass, and it seems maybe your changes have broken the Swift and C++ tests?
I've reviewed your changes, so I'm confident this is really minor.

File: runtime-testsuite/resources/org/antlr/v4/test/runtime/templates/Swift.test.stg (hunk 1 above)
Comment (marcospassos) [reply runtime-testsuite/resources/org/antlr/v4/test/runtime/templates/Swift.test.stg]: it certainly escaped me

**Comment by parrt** [comment]
Cool. I'll wait to see what the CI builds do

**Comment by ericvergnaud** [comment]
@marcospassos 
we're almost there! Travis CI is now green.
AppVeyor isn't. Can you have a look?

**Comment by marcospassos** [comment]
@ericvergnaud, thank you for reviewing.

@parrt I've fixed all remaining issues, and now all tests are passing.

**Comment by parrt** [comment]
Wow! We are all green!

So for release purposes, what should i add to my human script? 

1. jump into php submodule and say `git submodule update`
2. do i need to push anywhere like I do for python, java?

**Comment by marcospassos** [comment]
You don't need to do anything :) 

Whenever we change something in the PHP runtime, we'll also submit a PR updating the submodule, so you don't have to spend your limited time doing it on every new release.

**Comment by parrt** [comment]
Ok, great! and heeeeeere we go!!!!

**Comment by marcospassos** [comment]
Thank you for all support, @parrt  and @ericvergnaud.

**Comment by amorimjuliana** [comment]
🎊 🎉💃

**Comment by parrt** [comment]
Thank you for the hard work!!!   (And thanks @ericvergnaud for reviewing!)

**Comment by renanBritz** [comment]
Thank you for your hard work!

I couldn't find info about ANTLR release cycles... so when do you guys plan to create a release with these improvements?

Best Regards!

**Comment by yar3333** [comment]
@renanBritz If you need antlr4+php right now, you can use my release: https://github.com/yar3333/antlr4-php-runtime

**Comment by marcospassos** [comment]
I'd not recommend that, as the resulting parser will not be compatible with the official version, nor with the official runtime. If you need to it now, the recommended way is to clone the repository and build it locally, which is [pretty simple](https://github.com/antlr/antlr4/blob/master/doc/building-antlr.md#get-the-source).

**Comment by renanBritz** [comment]
Thanks for the reply. I'll try that.

**Pull request merged**